### DISCUSSION TITLE
BladeRF support

### DISCUSSION THREAD

**Discussion by improperaffo** [srsran/srsRAN_Project#222] [General Help] [answered]
Dear all,
Is there a plan for the support of BladeRFs in srsRAN_Project?

I see that the ZeroMQ drivers are already implemented, has anyone successfully used a BladeRF through ZMQ?

Thanks

**Comment by andrepuschmann** [comment]
There is currently no native support for anything else than OFH, UHD devices and ZMQ. However, there is a Soapy Wrapper for UHD. So any device that is available through Soapy can be used with UHD. With this you could try the BladeRF. I am not sure how usable this is though.

**Reply by improperaffo** [reply]
Thank you for the tip, we'll look into it.

**Comment by improperaffo** [comment, marked as answer]
Update:
we were able to run srsRAN using a BladeRF xA9 via the SoapySDR API and COTS UEs could connect to it. The main things we had to do were

- Install [SoapySDR](https://github.com/pothosware/SoapySDR), [SoapyBladeRF](https://github.com/pothosware/SoapyBladeRF), [SoapyUHD](https://github.com/pothosware/SoapyUHD) and [BladeRF](https://github.com/Nuand/bladeRF)
- Remove the reference to `time_source` in `lib/radio/uhd/radio_uhd_device.h:254`
- Set the tx buffer timeout (`TRANSMIT_TIMEOUT_S`) to 0 in `srsRAN_Project/lib/radio/uhd/radio_uhd_tx_stream.h:44`
- Set the rx buffer timeout (`RECEIVE_TIMEOUT_S`) to 0 in `srsRAN_Project/lib/radio/uhd/radio_uhd_rx_stream.h:40`

Versions used:

- FPGA image: `0.15.0`
- Firmware: `2.4.0`
- libbladeRF: `2.5.0` 
- UHD: `4.5.0.0`
- Soapy Lib: `0.8.1-1`


Update 2:
In order to avoid the following error:
```
[ERROR @ host/libraries/libbladeRF/src/backend/usb/libusb.c:1089] Transfer timed out for buffer 0x56349a19be30
```
we also coded a reset of the device in the destructor of `bladeRF_SoapySDR` in `bladeRF_Settings.cpp` as such:
```
bladeRF_SoapySDR::~bladeRF_SoapySDR(void)
{
    SoapySDR::logf(SOAPY_SDR_INFO, "bladerf_device_reset()");
    if (_dev != NULL) {
        bladerf_device_reset(_dev);
        // bladerf_close(_dev);
    }
}
```

Plus we changed the gain setting function in the SoapyBladeRF library: in `bladeRF_Settings.cpp`, instead of using `bladerf_set_gain_stage`, we are using `bladerf_set_gain`. Note that this requires to build the deb package again.

**Reply by andrepuschmann** [reply]
Excellent - thanks for reporting @improperaffo - I'll see if we can make the required changes run-time configurable so source changes aren't needed.
Did you experience any performance overhead compared to a e.g. USRP running nattily under UHD?

**Reply by improperaffo** [reply]
We did some testing and the Soapy layer does not introduce much latency. Here is a CDF of some ping commands to a nearby server using `speedtest-cli`.
![ping_soapyvsuhd](https://github.com/srsran/srsRAN_Project/assets/96285590/c7a242b1-dd62-4eda-8f04-3efa1cec4c6f)
We could not really compare the throughput, as it is hard to match the gains of the two boards.

**Reply by davwheat** [reply]
These changes allowed me to start the gNB, but I get errors when it actually begins trying to transmit anything:

```
Connecting to AMF on 10.30.3.10:38412
Available radio types: uhd.
[INFO] [UHD] linux; GNU C++ version 11.4.0; Boost_107400; UHD_4.6.0.0-3-g080b1baa
[INFO] [LOGGING] Fastpath logging disabled at runtime.
No device type given, found device with address 'backend=libusb,device=0x03:0x02,driver=bladerf,instance=0,label=BladeRF #0 [4c4ceb8c..4e613cf3],serial=4c4ceb8c96ee47e6977a37dc4e613cf3,type=soapy'
Making USRP object with args 'biastee_tx=true,type=soapy'
[INFO] [UHDSoapyDevice] bladerf_open_with_devinfo()
[INFO] [UHDSoapyDevice] bladerf_get_serial() = 4c4ceb8c96ee47e6977a37dc4e613cf3
[INFO] [UHDSoapyDevice] setSampleRate(Rx, 0, 4.000000 MHz), actual = 4.000000 MHz
[INFO] [UHDSoapyDevice] setSampleRate(Tx, 0, 4.000000 MHz), actual = 4.000000 MHz
[INFO] [UHDSoapyDevice] setSampleRate(Tx, 0, 19.200000 MHz), actual = 19.200000 MHz
[INFO] [UHDSoapyDevice] setSampleRate(Tx, 1, 19.200000 MHz), actual = 19.200000 MHz
[INFO] [UHDSoapyDevice] setSampleRate(Rx, 0, 19.200000 MHz), actual = 19.200000 MHz
[INFO] [UHDSoapyDevice] setSampleRate(Rx, 1, 19.200000 MHz), actual = 19.200000 MHz
Cell pci=1, bw=5 MHz, dl_arfcn=375500 (n3), dl_freq=1877.5 MHz, dl_ssb_arfcn=375410, ul_freq=1782.5 MHz

==== gNodeB started ===
Type <t> to view trace
[ERROR @ host/libraries/libbladeRF/src/backend/usb/libusb.c:1089] Transfer timed out for buffer 0x56349a19be30
[ERROR @ host/libraries/libbladeRF/src/backend/usb/libusb.c:1089] Transfer timed out for buffer 0x56349a19fe40
[ERROR @ host/libraries/libbladeRF/src/backend/usb/libusb.c:1089] Transfer timed out for buffer 0x56349a1a3e50
[ERROR @ host/libraries/libbladeRF/src/backend/usb/libusb.c:1089] Transfer timed out for buffer 0x56349a1a7e60
[ERROR @ host/libraries/libbladeRF/src/backend/usb/libusb.c:1089] Transfer timed out for buffer 0x56349a1abe70
```

Googling this, it seems to be a pretty common issue with bladeRF's and Nuand never seem to acknowledge the issue, despite several posts about it on their forum and multiple GitHub issues.

I guess I'm out of luck (and $500 or so).

**Reply by improperaffo** [reply]
Hi David, just to confirm, you did set the tx and rx buffer timeout values to 0 in the code?

**Reply by davwheat** [reply]
Yes, I did. I've since found that I can work around this *sometimes* by unplugging and plugging the bladeRF back in to the computer and then starting the gNB. I have to do this every time I stop the gNB, and it doesn't always work.

**Reply by improperaffo** [reply]
Oh right, I forgot about that. We just coded a reboot command to the BladeRF when stopping the gNB, so that we did not have to unplug it and replug it

**Reply by platini2** [reply]
@improperaffo , is the reset working for you? i tried to change those lines but the bladeRF is not resetting and i get the buffer time out.

**Reply by HaoxinSEU** [reply]
Hi @platini2 ,

Yes, the reset function works for us. We have:
```
bladeRF_SoapySDR::~bladeRF_SoapySDR(void)
{
    if (_dev != NULL) {
        bladerf_device_reset(_dev);
    }
}
```

**Comment by maxbraunschmied** [comment]
Hello,

I got my bladeRF xA4 to work with srsRAN using the tips mentioned in this thread.
However, my tx-gain seems to be out of range. Any ideas what values should be used in the conf.yaml file?

```
Connecting to AMF on 127.0.1.100:38412
Available radio types: uhd.

[INFO] [UHD] linux; GNU C++ version 12.2.0; Boost_107400; UHD_4.6.0.0-3-g080b1baa
[INFO] [LOGGING] Fastpath logging disabled at runtime.
No device type given, found device with address 'backend=libusb,device=0x02:0x03,driver=bladerf,instance=0,label=BladeRF #0 [da50ad69..b6c5a6f6],serial=da50ad69a8c1448b9af85e7bb6c5a6f6,type=soapy'
Making USRP object with args 'type=soapy'
[INFO] [UHDSoapyDevice] bladerf_open_with_devinfo()
[INFO] [UHDSoapyDevice] bladerf_get_serial() = da50ad69a8c1448b9af85e7bb6c5a6f6
[INFO] [UHDSoapyDevice] setSampleRate(Rx, 0, 4.000000 MHz), actual = 4.000000 MHz
[INFO] [UHDSoapyDevice] setSampleRate(Tx, 0, 4.000000 MHz), actual = 4.000000 MHz
[INFO] [UHDSoapyDevice] setSampleRate(Tx, 0, 23.040000 MHz), actual = 23.040000 MHz
[INFO] [UHDSoapyDevice] setSampleRate(Tx, 1, 23.040000 MHz), actual = 23.040000 MHz
[INFO] [UHDSoapyDevice] setSampleRate(Rx, 0, 23.040000 MHz), actual = 23.040000 MHz
[INFO] [UHDSoapyDevice] setSampleRate(Rx, 1, 23.040000 MHz), actual = 23.040000 MHz
Error: setting gain for transmitter 0. Tx gain (i.e., 60.0 dB) is out-of-range. Range is [-89.75000762939453, 0.0] dB in steps of 0.25 dB.
```

**Reply by improperaffo** [reply]
Hi Max. If I recall correctly, the gain for the BladeRF using soapy is a bit mangled. The Tx goes from -89 to 0 (-89 being the highest). While the Rx should be the classic 0 to 66.

**Reply by improperaffo** [reply]
Hi. We were referring to the `lib/` folder inside the project directory (i.e., `srsRAN_Project/lib/radio/uhd/radio_uhd_device.h:254`)

**Reply by davwheat** [reply]
Out of interest, what values for Rx/Tx gain have you all found best success with? I end up with something like 75% UL BLER no matter what I've tried right now (and I'm open for any suggestions on how to improve it!).

**Reply by HaoxinSEU** [reply]
Hi @davwheat ,

We changed the gain setting function in the library _SoapyBladeRF_.
 
In `bladeRF_Settings.cpp`, instead of using `bladerf_set_gain_stage`, we change to use `bladerf_set_gain`. Then we build the deb package again.

Because in the beginning we noticed that the gain setting seems not working, and after changing it, we can clearly see it works and changes the output power.

Hope this helps.

**Reply by platini2** [reply]
@HaoxinSEU , Can you please clarify what lines you changed on bladeRF_Settings.cpp to fix the gain function?

is it this? 

sed -i '373s/ret =.*/ret = bladerf_set_gain(_dev, _toch(direction, channel), bladerf_gain(std::round(value)));/g' ./bladeRF_Settings.cpp

**Reply by HaoxinSEU** [reply]
Hi,

In the function 
```
void bladeRF_SoapySDR::setGain(const int direction, const size_t channel, const std::string &name, const double value)
```
we change:
```
int ret = bladerf_set_gain_stage(_dev, _toch(direction, channel), name.c_str(), bladerf_gain(std::round(value)));
```
to:
```
ret = bladerf_set_gain(_dev, _toch(direction, channel), bladerf_gain(std::round(converted_value)));
```

**Comment by MamdouhEgy** [comment]
(too long comment sorry for this)

Could anyone help me figuring out why i cannot see any devices:
I use ubuntu 20.04 & BladeRF xA4 (connected and 4 lamps are on) and i followed the above instructions to install [SoapySDR](https://github.com/pothosware/SoapySDR), [SoapyBladeRF](https://github.com/pothosware/SoapyBladeRF), [SoapyUHD](https://github.com/pothosware/SoapyUHD) and [BladeRF](https://github.com/Nuand/bladeRF), removed the time source reference and changed the tx&rx values as following:
**removed the time source reference:** 
#if UHD_VERSION < 3140099
    return safe_execution([this, &sync_src, &clock_src]() {
      std::vector<std::string> clock_sources = usrp->get_clock_sources(0);
      if (std::find(clock_sources.begin(), clock_sources.end(), clock_src) == clock_sources.end()) {
        on_error("Invalid clock source {}. Supported: {}", clock_src, span<const std::string>(clock_sources));
        return;
      }
      usrp->set_clock_source(clock_src);
    });

**changed the tx&rx values as following:**
  static constexpr double TRANSMIT_TIMEOUT_S = 0;
  static constexpr double RECEIVE_TIMEOUT_S = 0f;
  
  **i executed the following commands in order and the o/p in the pic:** 
bladeRF-cli -e 'version'   
SoapySDRUtil --info 
uhd_usrp_probe
lsusb 
sudo ./gnb -c gnb_rf_b200_tdd_n78_20mhz.yml.
  
![commands1](https://github.com/srsran/srsRAN_Project/assets/42903127/fb096330-fc09-4de0-8b91-12ba4538743e)


  **I always get as shown above:**
Error: no radio devices found.
Failed to open device with address '': 
srsGNB ERROR: Unable to create radio session.


**ps aux | grep open***
![psaux](https://github.com/srsran/srsRAN_Project/assets/42903127/c074eed3-4417-4c0d-ace5-e28fd695fe84)

**the amf.yaml file:**
![amf](https://github.com/srsran/srsRAN_Project/assets/42903127/2e7d5948-c405-4c64-9daf-4ec0618bab4f)

**the upf.yaml file:**
![upf](https://github.com/srsran/srsRAN_Project/assets/42903127/e40b104c-a792-4c41-b9bc-f08f9e140f5e)

**the gnb_rf_b200_tdd_n78_20mhz.yml file:**
![gnb](https://github.com/srsran/srsRAN_Project/assets/42903127/f87687e9-a7eb-4547-8803-e0e668f5b3c9)

**tail -f /var/log/open5gs/amf.log**
![log](https://github.com/srsran/srsRAN_Project/assets/42903127/91ea9337-0ce3-4430-9c48-cec2d8037bab)

**apt list --installed | grep uhd**
![uhd](https://github.com/srsran/srsRAN_Project/assets/42903127/70a468c6-12ac-49da-81a5-cd23047d5688)

**ip address | grep -A 5 "^lo:"**
![loopinterface](https://github.com/srsran/srsRAN_Project/assets/42903127/759c8c0a-7e5b-4ef6-8dde-cc8e01ba4ae2)

**Reply by HaoxinSEU** [reply]
Hi,

I'm not 100% sure what's the problem. But can you give it a try with UHD 4.5.0 instead of 4.6.0? I remember at some point the library version played a role. So try to remove other versions of libhud, and just install libuhd4.5.0, then redo the previous steps (e.g., build SoapyUHD). 

After this, try to run `uhd_usrp_probe` to check if the device appears.

If this doesn't solve the problem, we can discuss in more details.

**Reply by MamdouhEgy** [reply]
Thanks @HaoxinSEU 

I did all again with **UHD 4.5.0.0** but as shown, still **No UHD Devices Found**

![Screenshot at 13-26-05](https://github.com/srsran/srsRAN_Project/assets/42903127/d05f2bb5-5ec8-410e-8094-b0081b573d7f)

**Reply by HaoxinSEU** [reply]
Can you run `apt list --installed | grep soapy` to see which packages you installed? Thanks

**Reply by MamdouhEgy** [reply]
thanks
Here for soapy and uhd, i noticed that uhs is depicted here with version 4.6.0 and 3.15:

![Screenshot at 13-40-56](https://github.com/srsran/srsRAN_Project/assets/42903127/e692204c-821f-4948-9e6a-b734677f274f)

**Reply by improperaffo** [reply]
Hi Mamdouh, you can try removing libuhd3.15.0 with `apt remove -y libuhd3.15.0` and then re install the Soapy libraries (`uhd-soapysdr`, `soapysdr0.7-module-bladerf`).

**Reply by MamdouhEgy** [reply]
i did that now and as you see, installing **uhd-soapysdr** is reinstalling **libuhd3.15.0**

![Screenshot at 14-25-28](https://github.com/srsran/srsRAN_Project/assets/42903127/1a19faf5-b34d-4455-8727-d9a217e2a89c)

**Reply by HaoxinSEU** [reply]
Hi,

Can you manually remove this `libuhd3.15.0`, after installing `uhd-soapysdr`?

**Reply by MamdouhEgy** [reply]
Hi,

i used **dpkg-query -L libuhd3.15.0** command then i deleted /usr/lib/x86_64-linux-gnu/**libuhd.so.3.15.0** file.

I think there is somehow conflict of the results of different commands as shown:

![Screenshot at 15-11-50](https://github.com/srsran/srsRAN_Project/assets/42903127/d3fc9cc4-2c16-4742-998c-ab118858a0a9)

**Reply by MamdouhEgy** [reply]
i install libuhd4.5.0 again using **sudo apt install libuhd4.5.0** as i installed it in the past using the cmake from the branch uhd4.5 here https://github.com/EttusResearch/uhd/tree/UHD-4.5.

now it is shown as the following:
![Screenshot at 15-21-03](https://github.com/srsran/srsRAN_Project/assets/42903127/4177e326-a70b-469c-a841-99a9d8d51f28)

**Reply by platini2** [reply]
My two cents, wthat could happen beacuse the soapyUHD is not the the UHD module path

try this before running uhd_find

export UHD_MODULE_PATH=/usr/lib/uhd/modules/

**Reply by bcrobin03** [reply]
Hi @MamdouhEgy, I am having the exact same problem as you, did you find a fix for it ?

**Reply by MamdouhEgy** [reply]
@bcrobin03 nothing worked and i was in a long vacation... i will start again... iam using 2 dockers now.. the first is  [srs-gnb-blade)](https://hub.docker.com/r/pavonis/srs-gnb-blade) and the second is open5gs container and i will try to connect both.

**Reply by Havok-Alwin** [reply]
any update here @MamdouhEgy
